src/spatialdata/dataloader/graph_loader.py

```diff
@@ -29,104 +46,90 @@ def build_graph(  # TODO: add rings for grids
-    if use_centroids:
-        gdf = gdf.centroid
-    # only one of k, threshold, or max_distance can be specified, the one corresponding to the method selected
+
+    if max_distance is not None:
+        assert max_distance > 0
+    if percentile is not None:
+        assert percentile >= 0
+        assert percentile <= 100
+
     if method == "knn":
         assert k is not None
         assert k > 0
-        assert max_distance is None or max_distance > 0
-        if not use_centroids:
-            logger.warning("method knn is only available for centroids, automatically set use_centroids=True.")
-        gdf_t = gdf.centroid
-        centroids = np.array(gdf_t.apply(lambda g: [g.x, g.y]).tolist())
+        assert not (percentile is not None and max_distance is not None)
+    if method == "radius":
+        assert k is None
+        assert (percentile is None) ^ (max_distance is None)
+    if method == "expansion":
+        assert k is None
+        assert percentile is None
+
+    if method in ["knn", "radius"] and not use_centroids:
+        logger.warning(f"method {method} only available for centroids, automatically set use_centroids=True.")
+        use_centroids = True
+
+    if use_centroids:
+        gdf_c = gdf.centroid
+
+    if method == "knn":
+        centroids = np.array(gdf_c.apply(lambda g: [g.x, g.y]).tolist())
         tree = KDTree(centroids)
 
         if max_distance is not None:
-            if percentile is not None:
-                logger.warning("both percentile and max_distance are set, will only use max_distance.")
+            d_tree, idx = tree.query(centroids, k=k, distance_upper_bound=max_distance)
+        elif percentile is not None:
+            assert percentile is not None
+            max_distance = _compute_percentile(tree, percentile, centroids)
             d_tree, idx = tree.query(centroids, k=k, distance_upper_bound=max_distance)
         else:
-            if percentile is not None:
-                logger.warning("computing percentile could take long for large datasets.")
-                centroids = np.array(centroids)
-                dd = tree.sparse_distance_matrix(
-                    tree,
-                    max_distance=np.sqrt(
-                        (np.max(centroids[:, 0]) - np.min(centroids[:, 0])) ** 2
-                        + (np.max(centroids[:, 1]) - np.min(centroids[:, 1])) ** 2
-                    ),
-                ).toarray()
-                np.fill_diagonal(dd, np.nan)
-                dd = dd[~np.isnan(dd)].reshape(dd.shape[0], dd.shape[1] - 1)
-                max_distance = np.percentile(dd, percentile)
-                d_tree, idx = tree.query(centroids, k=k, distance_upper_bound=max_distance)
-            else:
-                d_tree, idx = tree.query(centroids, k=k)
+            d_tree, idx = tree.query(centroids, k=k)
 
         idx = idx[:, 1:]
         if idx.size == 0:
-            return
+            return np.array([[]])
```

Comment:
The return type was wrong. The function is supposed to return an array but a `None` was being returned.